File: packages/battle_engine/lib/src/battle_state.dart
import 'active_status.dart';
import 'combatant.dart';
import 'field_effect.dart';
import 'hp_pool.dart';
import 'status_effect.dart';

/// Snapshot of an in-progress 1v1 battle: the two participants, any active
/// field effects (produced by element combinations or ability mutations),
/// the status effects and HP of each combatant, whose turn it is, and who
/// won (if the battle is over).
///
/// Pure data, immutable. Turn order, damage resolution and win/loss belong
/// to [TurnEngine] — this class only represents a valid state and how to
/// derive the next one.
class BattleState {
  final Combatant playerA;
  final Combatant playerB;
  final Combatant currentTurn;
  final List<FieldEffect> activeFieldEffects;
  final Map<Combatant, List<ActiveStatus>> combatantStatuses;
  final Map<Combatant, HpPool> hp;
  final Combatant? winner;

  BattleState({
    required this.playerA,
    required this.playerB,
    required this.currentTurn,
    List<FieldEffect> activeFieldEffects = const [],
    Map<Combatant, List<ActiveStatus>>? combatantStatuses,
    Map<Combatant, HpPool>? hp,
    this.winner,
  })  : activeFieldEffects = List.unmodifiable(activeFieldEffects),
        combatantStatuses = Map<Combatant, List<ActiveStatus>>.unmodifiable({
          playerA: List<ActiveStatus>.unmodifiable(
            combatantStatuses?[playerA] ?? const <ActiveStatus>[],
          ),
          playerB: List<ActiveStatus>.unmodifiable(
            combatantStatuses?[playerB] ?? const <ActiveStatus>[],
          ),
        }),
        hp = Map<Combatant, HpPool>.unmodifiable({
          playerA: hp?[playerA] ?? const HpPool(max: 100, current: 100),
          playerB: hp?[playerB] ?? const HpPool(max: 100, current: 100),
        }) {
    if (playerA == playerB) {
      throw ArgumentError('playerA and playerB must be distinct combatants');
    }
    if (currentTurn != playerA && currentTurn != playerB) {
      throw ArgumentError('currentTurn must be playerA or playerB');
    }
  }

  /// Starts a new battle with an empty field, no statuses, [playerA]
  /// acting first. [playerAMaxHp]/[playerBMaxHp] default to 100 — a
  /// convenience for callers that don't care about HP (most tests). Real
  /// gameplay always computes and passes the real value (base + Skill
  /// Tree bonus) explicitly; nothing in production code relies on this
  /// default.
  factory BattleState.start({
    required Combatant playerA,
    required Combatant playerB,
    int playerAMaxHp = 100,
    int playerBMaxHp = 100,
  }) {
    return BattleState(
      playerA: playerA,
      playerB: playerB,
      currentTurn: playerA,
      hp: {
        playerA: HpPool(max: playerAMaxHp, current: playerAMaxHp),
        playerB: HpPool(max: playerBMaxHp, current: playerBMaxHp),
      },
    );
  }

  BattleState copyWith({
    Combatant? currentTurn,
    List<FieldEffect>? activeFieldEffects,
    Map<Combatant, List<ActiveStatus>>? combatantStatuses,
    Map<Combatant, HpPool>? hp,
    Combatant? winner,
  }) {
    return BattleState(
      playerA: playerA,
      playerB: playerB,
      currentTurn: currentTurn ?? this.currentTurn,
      activeFieldEffects: activeFieldEffects ?? this.activeFieldEffects,
      combatantStatuses: combatantStatuses ?? this.combatantStatuses,
      hp: hp ?? this.hp,
      winner: winner ?? this.winner,
    );
  }

  /// Returns a new state with [effect] added to the field.
  BattleState withFieldEffect(FieldEffect effect) {
    return copyWith(activeFieldEffects: [...activeFieldEffects, effect]);
  }

  /// Returns the other participant relative to [combatant].
  Combatant opponentOf(Combatant combatant) {
    if (combatant == playerA) return playerB;
    if (combatant == playerB) return playerA;
    throw ArgumentError.value(
      combatant,
      'combatant',
      'is not part of this battle',
    );
  }

  /// Active statuses currently applied to [combatant].
  List<ActiveStatus> statusesOf(Combatant combatant) {
    _requireParticipant(combatant);
    return combatantStatuses[combatant] ?? const [];
  }

  /// Whether [combatant] currently has [effect] applied.
  bool hasStatus(Combatant combatant, StatusEffect effect) {
    return statusesOf(combatant).any((status) => status.effect == effect);
  }

  /// Returns a new state with [status] applied to [target].
  BattleState withStatusApplied(Combatant target, ActiveStatus status) {
    _requireParticipant(target);
    final updated = Map<Combatant, List<ActiveStatus>>.from(
      combatantStatuses,
    );
    updated[target] = [...statusesOf(target), status];
    return copyWith(combatantStatuses: updated);
  }

  /// Returns a new state with every active instance of [effect] removed
  /// from [target].
  BattleState withStatusRemoved(Combatant target, StatusEffect effect) {
    _requireParticipant(target);
    final updated = Map<Combatant, List<ActiveStatus>>.from(
      combatantStatuses,
    );
    updated[target] =
        statusesOf(target).where((status) => status.effect != effect).toList();
    return copyWith(combatantStatuses: updated);
  }

  /// Ticks every active status for both combatants by one turn, removing
  /// any that expire. Callers decide when this should run.
  BattleState withStatusesTicked() {
    final updated = <Combatant, List<ActiveStatus>>{
      for (final combatant in [playerA, playerB])
        combatant: statusesOf(combatant)
            .map((status) => status.tick())
            .where((status) => !status.isExpired)
            .toList(),
    };
    return copyWith(combatantStatuses: updated);
  }

  /// HP pool of [combatant].
  HpPool hpOf(Combatant combatant) {
    _requireParticipant(combatant);
    return hp[combatant]!;
  }

  /// Returns a new state with [amount] of damage applied to [target]. If
  /// this brings [target] to 0 HP and nobody has won yet, [target]'s
  /// opponent becomes the winner. Never un-sets an existing winner.
  BattleState withDamage(Combatant target, int amount) {
    _requireParticipant(target);
    final updatedPool = hpOf(target).withDamage(amount);
    final updatedHp = Map<Combatant, HpPool>.from(hp)..[target] = updatedPool;
    final newWinner = (winner == null && updatedPool.isDefeated)
        ? opponentOf(target)
        : winner;
    return copyWith(hp: updatedHp, winner: newWinner);
  }

  /// Returns a new state with [target]'s max (and current) HP increased by
  /// [amount] — used when a mid-battle bonus (e.g. a Skill Tree node)
  /// raises the ceiling.
  BattleState withMaxHpIncreased(Combatant target, int amount) {
    _requireParticipant(target);
    final updatedHp = Map<Combatant, HpPool>.from(hp)
      ..[target] = hpOf(target).withMaxIncreased(amount);
    return copyWith(hp: updatedHp);
  }

  void _requireParticipant(Combatant combatant) {
    if (combatant != playerA && combatant != playerB) {
      throw ArgumentError.value(
        combatant,
        'combatant',
        'is not part of this battle',
      );
    }
  }
}

File: backend/src/battle-rules/battle-state.ts
import { tick as tickStatus, isExpired } from "./active-status.js";
import {
  withDamage as poolWithDamage,
  withMaxIncreased as poolWithMaxIncreased,
  isDefeated,
} from "./hp-pool.js";
import { TurnValidationError } from "./errors.js";
import type { ActiveStatus, BattleState, FieldEffect, HpPool } from "./types.js";

/** Builds a validated BattleState — mirrors the invariants of
 * BattleState's constructor in battle_engine. `hp` defaults each player to
 * 100/100 (mirrors BattleState's plain constructor default in Dart — there
 * is no Skill Tree here to raise it); `combatantStatuses` defaults to no
 * active statuses for either player. */
export function createBattleState(input: {
  playerAId: string;
  playerBId: string;
  currentTurnId: string;
  activeFieldEffects?: readonly FieldEffect[];
  hp?: Readonly<Record<string, HpPool>>;
  combatantStatuses?: Readonly<Record<string, readonly ActiveStatus[]>>;
  winner?: string | null;
}): BattleState {
  if (input.playerAId === input.playerBId) {
    throw new TurnValidationError(
      "playerAId and playerBId must be distinct",
    );
  }
  if (
    input.currentTurnId !== input.playerAId &&
    input.currentTurnId !== input.playerBId
  ) {
    throw new TurnValidationError(
      "currentTurnId must be playerAId or playerBId",
    );
  }

  return {
    playerAId: input.playerAId,
    playerBId: input.playerBId,
    currentTurnId: input.currentTurnId,
    activeFieldEffects: input.activeFieldEffects ?? [],
    hp: {
      [input.playerAId]: input.hp?.[input.playerAId] ?? { max: 100, current: 100 },
      [input.playerBId]: input.hp?.[input.playerBId] ?? { max: 100, current: 100 },
    },
    combatantStatuses: {
      [input.playerAId]: input.combatantStatuses?.[input.playerAId] ?? [],
      [input.playerBId]: input.combatantStatuses?.[input.playerBId] ?? [],
    },
    winner: input.winner ?? null,
  };
}

/** Mirrors BattleState.opponentOf. */
export function opponentOf(state: BattleState, combatantId: string): string {
  if (combatantId === state.playerAId) return state.playerBId;
  if (combatantId === state.playerBId) return state.playerAId;
  throw new TurnValidationError(
    `"${combatantId}" is not part of this battle`,
  );
}

/** HP pool of a combatant. */
export function hpOf(state: BattleState, combatantId: string): HpPool {
  const pool = state.hp[combatantId];
  if (!pool) {
    throw new TurnValidationError(
      `"${combatantId}" is not part of this battle`,
    );
  }
  return pool;
}

/** Mirrors BattleState.withDamage: applies `amount` of damage to `targetId`
 * and, if that brings them to 0 HP and nobody has won yet, sets their
 * opponent as the winner. Never un-sets an existing winner. */
export function withDamage(
  state: BattleState,
  targetId: string,
  amount: number,
): BattleState {
  const updatedPool = poolWithDamage(hpOf(state, targetId), amount);
  const newWinner =
    state.winner === null && isDefeated(updatedPool)
      ? opponentOf(state, targetId)
      : state.winner;

  return {
    ...state,
    hp: { ...state.hp, [targetId]: updatedPool },
    winner: newWinner,
  };
}

/** Returns a new state with `targetId`'s max (and current) HP increased by
 * `amount` — used when a mid-battle bonus (e.g. a Skill Tree node) raises
 * the ceiling. */
export function withMaxHpIncreased(
  state: BattleState,
  targetId: string,
  amount: number,
): BattleState {
  return {
    ...state,
    hp: { ...state.hp, [targetId]: poolWithMaxIncreased(hpOf(state, targetId), amount) },
  };
}

/** Active statuses currently applied to a combatant. */
export function statusesOf(
  state: BattleState,
  combatantId: string,
): readonly ActiveStatus[] {
  return state.combatantStatuses[combatantId] ?? [];
}

/** Whether a combatant currently has a status with this `effectId`. */
export function hasStatus(
  state: BattleState,
  combatantId: string,
  effectId: string,
): boolean {
  return statusesOf(state, combatantId).some((s) => s.effectId === effectId);
}

/** Returns a new state with `status` applied to `targetId`. */
export function withStatusApplied(
  state: BattleState,
  targetId: string,
  status: ActiveStatus,
): BattleState {
  return {
    ...state,
    combatantStatuses: {
      ...state.combatantStatuses,
      [targetId]: [...statusesOf(state, targetId), status],
    },
  };
}

/** Returns a new state with every active instance of `effectId` removed
 * from `targetId`. */
export function withStatusRemoved(
  state: BattleState,
  targetId: string,
  effectId: string,
): BattleState {
  return {
    ...state,
    combatantStatuses: {
      ...state.combatantStatuses,
      [targetId]: statusesOf(state, targetId).filter((s) => s.effectId !== effectId),
    },
  };
}

/** Ticks every active status for both combatants by one turn, removing
 * any that expire. */
export function withStatusesTicked(state: BattleState): BattleState {
  const updated: Record<string, readonly ActiveStatus[]> = {};
  for (const combatantId of [state.playerAId, state.playerBId]) {
    updated[combatantId] = statusesOf(state, combatantId)
      .map(tickStatus)
      .filter((s) => !isExpired(s));
  }
  return { ...state, combatantStatuses: updated };
}
